src/nvim/event/process.c

```diff
@@ -223,11 +223,13 @@ void process_stop(Process *proc) FUNC_ATTR_NONNULL_ALL
 
   switch (proc->type) {
     case kProcessTypeUv:
+      os_proc_tree_kill(proc->pid, SIGTERM);
       // Close the process's stdin. If the process doesn't close its own
       // stdout/stderr, they will be closed when it exits(possibly due to being
-      // terminated after a timeout)
+      // terminated after a timeout).
+      // This should happen only after sending SIGTERM, for consistent behavior
+      // with e.g. "cat -".
       stream_may_close(&proc->in);
-      os_proc_tree_kill(proc->pid, SIGTERM);
       break;
```

Comment:
Previously this would be flaky:

> while true; do TEST_FILTER="jobs avoids sending final newline" test/run.sh test/functional/core/job_spec.lua || break; done

Reply:
Taken out into https://github.com/neovim/neovim/pull/10584 already.